0015-ilp-addresses/0015-ilp-addresses.md

```diff
@@ -0,0 +1,179 @@
+
+- If the matching address is marked for local delivery, the connector prepares a transfer to that address in one of the ledgers connected to it. The connector maps the ILP address to an account within the ledger. (This is the base case.)
+- If the matching address is marked as forwarded delivery, it has the address of another connector associated with it in the routing table. The connector makes a routing lookup on the connector associated with the address. (This is the recursive case.)
+
+## Address Requirements
+
+ILP Addresses must meet the following requirements:
+
+1. The address MUST begin with a prefix indicating the allocation scheme. See [Allocation Schemes](#allocation-schemes) for more information.
+2. Each "segment" of the address MUST contain one or more of the following characters:
+    - Alphanumeric characters, upper or lower case. (Addresses are **case-sensitive** so that they can contain data encoded in formats such as base64url.)
+    - Underscore (`_`)
+    - Tilde (`~`)
+    - Hyphen (`-`)
+3. Each segment MUST be separated from other segments by a period character (`.`).
+4. Address prefixes MUST end in a period (`.`) character and MAY contain any number of segments after the allocation scheme prefix.
+5. Destination addresses MUST NOT end in a period (`.`) character, and MUST contain at least two segments after the allocation scheme prefix.
+6. The total length of an ILP Address must be no more than **1023 characters** including the allocation scheme prefix, separators, and all segments.
+
+The following ABNF specification defines the format for all ILP addresses and address prefixes:
+
+```abnf
+address     = scheme separator *prefix [segment]
+                    ; the last segment is REQUIRED for destination addresses
+
+scheme      = "g" / "private" / "example" / "peer" / "self" /
+              "test1" / "test2" / "test3"
+
+separator   = "."
+
+prefix      = 1*(segment separator)
+
+segment     = 1*( ALPHA / DIGIT / "_" / "~" / "-" )
+```
+
+You can also use the following regular expressions to verify the same requirements:
+
+| Address Type        | Regular Expression                                     |
+|:--------------------|:-------------------------------------------------------|
+| All addresses       | `^(g|private|example|peer|self|test[1-3])\.([a-zA-Z0-9_~-]+\.)*([a-zA-Z0-9_~-]+)?$` |
+| Address prefix      | `^(g|private|example|peer|self|test[1-3])\.([a-zA-Z0-9_~-]+\.)*$` |
+| Destination address | `^(g|private|example|peer|self|test[1-3])\.([a-zA-Z0-9_~-]+\.)+[a-zA-Z0-9_~-]+$` |
+
+(Note that neither the ABNF nor the regular expressions enforce the maximum length requirement. Remember to check that total length of the address is **1023 characters or less**.)
```

Comment:
This seems like it's asking for people to shoot themselves in the foot. Can't we just provide a regex that starts with `(?=^.{1,1023}$)`? The only issue with this is that Go's regexes don't understand lookaheads. Not sure about Java, but JS, Python, and PHP are all fine with that syntax.